### DISCUSSION TITLE
How to limit cpptools memory use

### DISCUSSION THREAD

**Discussion by PJH0x0** [microsoft/vscode-cpptools#9482] [General]
![image](https://user-images.githubusercontent.com/24682416/174752617-2e46268f-ca3f-4568-b129-e5bf1adc3c3f.png)
It is very bad memory use when I open android open source project with vscode

**Comment by sean-mcmanus** [comment]
It is not expected for the cpptools process to use a lot of memory under normal conditions. It's possible too many files/folders are getting indexed. You could try adding folders to the C_Cpp.files.exclude setting to scan less files for global symbols. If you set C_Cpp.loggingLevel to "Debug" the logging might indicate what processing is occurring that could be contributing to the memory usage. We have settings like C_Cpp.maxMemory you could try, but I don't believe that would help in your situation.

**Reply by madscientist42** [reply]
"It is not expected for the cpptools process to use a lot of memory under normal conditions. It's possible too many files/folders are getting indexed." 

That's quite an assumption there.  It's showing hubris.  What the heck is, "too many files/folders"?  Honestly, let's ask a slightly differing question of you instead...

What happens when you have a Yocto build system where you need both the Yocto bitbake extension and the C/C++ one because you're doing code dev like you're really supposed to using Yocto?  I can tell you.

It finds the literally thousands of project source files for each one of the tooling build-outs for the final image recipe and tries to index them with disastrous results.  The very thing you claim isn't normal conditions.  I'll own that it might not be, "normal," for what you're thinking of- but it's normal for Embedded Linux dev use.  It's the tool path and workflow that was developed for it by the Linux Foundation at this point in time.

My machine?  32 GiB of RAM.  64 GiB of NVMe Gen3 backed swap.  Your little tool has taken to curb stomping my box so I can't use it at-all for my workflows.  

There should be an explicitly documented means to nerf this as asked for.  There should, additionally, be actual limits, set somewhat low, that a user can override, **_also documented well_**, that keep it from curb-stomping systems.

I can see this extension curb-stomping low-spec systems under somewhat regular circumstances as I find at least part of the key settings that your code doesn't fully seem to honor to be set to eat the whole of the available system memory.  C'mon...seriously?

**Reply by sean-mcmanus** [reply]
@madscientist42 An example of "too many files" might be the root "/" drive (see https://github.com/microsoft/vscode-cpptools/issues/11993 ). You should add build output folders to the C_Cpp.files.exclude setting.

**Reply by bobbrow** [reply]
Hi @madscientist42,

I recognize your scenario from the Reviews page.  Would you like to schedule a call with me and we can review what's going wrong with your workspace?  We're not experts in Yocto and I'd like to understand how you're configured so we can make sure everyone using Yocto has a good experience.

This is our public support email address: c_cpp_support@microsoft.com. Write me a message there and we can coordinate a time that works for you.

Thanks,
Bob

**Comment by stellarpower** [comment]
It's not uncommon for me to see CPPtools using in excess of 30 gig... well, not "not uncommon", it always is, and I have several C++ projects open at once, so each one is using a load on its own. I've had to set a shell job to sit in a loop sleeping and then killing it repeatedly before, when my system is struggling for memory and I need the least important thing to die. 

I think one of the issues for me is - without knowing how it works beyond probably a language server - I feel like it's probably trying to work like a compiler and understand the whole project. Even the python extension often underlines things that are fine and fails to underline things that aren't, so gut feel is, it just won't, it's asking too much. When C++ project builds are much more complicated it can't possibly know all the flags and include directories, dependencies, that are needed to build what I'm working on right now. And that's completely fine. I don't know if there's a setting for this, but I'd personally like it to run in much more of a stripped-back version. I don't expect it to be able to find semantic errors and add red squiggly underlines, for that I'll build and control-click on the error message, which to be honest I need in full, with a backtrace of template instantiations, to diagnose the problem anyway. I'd like to be able to hit F4 to change header to source, something that sometimes takes a minute to run and then gets totally the wrong file. If it can find some basic typos and issues between one source file and one or two of the headers it includes directly, just in the scope that it cans see in the document I'm currently typing (e.g. using -> when it's an object on the stack that's not a pointer) then great, otherwise it appears to be using a massive amount of system resources by default trying to provide all the insights a compiler could except without knowing how my compiler is set up and failing in the process. I'd prefer more textual stuff like being able to refactor a method from a header into a source file than for it to attempt to catch subtle issues and parse an entire project. And I guess for it maybe to set a limit on how hard it's gonna work trying before it give up, as that's better than crashing my system or spinning up the fans massively, cause I see several mentions of others' experiencing issues like this, so I think it makes sense to have a hard limit by default.

**Reply by sean-mcmanus** [reply]
@stellarpower It sounds like you're hitting a bug. Is 30 GB with a single cpptools process or all of the cpptools-srv included? We can probably configure like the build does if you supply a compile_commands.json or use CMake Tools or Makefile Tools as the configurationProvider.

**Reply by stellarpower** [reply]
Afraid I'm not sure. I never trust the output the Linux kernel gives, it's currently saying code is using a terabyte of virtual when there is only a tenth of that physically installed and barely any swap so honestly I never know where to look to get any reasonable idea. I'm pretty sure I've seen both. Htoonshoukd be set up to collapse threads under the same process but I still usually see duplicates, especially with many applications now using a multiprocess rather than a multi-threaded architecture.

I haven't configured it myself before, it usually just starts in the background. I guess I might be able to generate compile commands as JSON. Often I am using several different build processes in the same window though, I usually just build through the shell.

**Reply by stellarpower** [reply]
I've not played with btop much, so I don;t know how it differs in what it reports to htop, but something like this is pretty common:

![image](https://github.com/user-attachments/assets/10a1b70c-d84a-49e6-8fac-e6fe3f0acbce)

**Reply by stellarpower** [reply]
Know it's not adding much in terms of helpful information, but just in terms of highlighting the severity of this - pretty sure cpptools brought down my desktop the other day. oomd ended up running and killed kwin. I have been keeping an eye on memory since, and kwin and firefox both very large so assumed they were the culprits - but I hadn't scrolled down far enough to see the multiple CPPtools running after that. I started my `killall cpptools` loop in the background and recovered 40-50G of RAM. So basically the combined effects of one intellisense tool is using as much memory as the entire rest of a multiuser system, with me logged in graphically with over 100 browser windows, 12 IDEs, tools like Blender, etc. all running. Since moving to a new installation I've noticed the whole system sometimes hangs for a few seconds. As long as it always comes back and doesn't crash, I don;t care, but am now wondering if it's cpptools causing stress in the background, so will see what happens in the next few days. 

tl;dr - on a brand new system with a new version of Code, this is at the level it's crashing or hanging a system with 128G installed, and losing me unsaved work and time trying to re-open everything I was just working on.

**Reply by bobbrow** [reply]
Thanks @stellarpower. My offer above to jump on a call to discuss this and try to understand the repro is extended to you as well if you're interested.

> This is our public support email address: [c_cpp_support@microsoft.com](mailto:c_cpp_support@microsoft.com). Write me a message there and we can coordinate a time that works for you.

IntelliSense is not supposed to eat that much memory and we check in on the process and kill it if it goes too high. However, there are limitations to our current implementation of that logic and it appears that you are hitting one of those limitations. That system may require a redesign...

**Reply by stellarpower** [reply]
Thanks, that's extremely generous. I just realised that there are settings I can tune in the options so maybe something there can help me. Even on my laptop I also found the cache used was like ten gig and I don't write that much code, so will have a play. And if I find time drop an email, appreciate it!

**Reply by sean-mcmanus** [reply]
@stellarpower I don't believe the existing performance settings are specifically designed to reduce excessive memory usage in the cpptools process -- other than C_Cpp.files.exclude. The memory usage by all the cpptools-srv processes can be affected by the various C_Cpp.max* performance settings though.

**Reply by stellarpower** [reply]
Agreed. But I should probably be making sure code is pointed towards the right tool for the environment I am using \;)

**Reply by dnwerner** [reply]
Maybe to add to this: This is an easy footgun for new users: We have a few dev-servers where cpptools managed to eat >240GB of memory as a path was seemingly misconfigured. This should not happen, as this can occur in changing work-envs.  Is there a way to potentially cut file traversal after a sufficiently high number of nodes - or add a pagination-like callback? In our env the only thing that finally stopped cpptools was oomd...

**Comment by tony-p** [comment]
Where can I find the documentation for the `C_Cpp.max*` settings?

* This link is dead https://code.visualstudio.com/docs/cpp/c-cpp-properties-schema-reference
   * archived versions don't mention it https://web.archive.org/web/20250428094658/https://code.visualstudio.com/docs/cpp/c-cpp-properties-schema-reference
* Can't find anything here https://code.visualstudio.com/docs/cpp/customize-cpp-settings
* This redirect doesn't mention them https://github.com/microsoft/vscode-cpptools/blob/main/Documentation/LanguageServer/Customizing%20Default%20Settings.md  

Note for our use case, we have a conda environment with >900 packages (which we typically reference for intellisense), but anytime we touch anything it generates 20Gb of cache data which seems overkill given (I think) it is referencing max 8Gb of files

**Reply by sean-mcmanus** [reply]
@tony-p The documentation is shown in the settings UI. C_Cpp settings are not in the c_cpp_properties.json so they wouldn't be in the schema docs. What exactly is 20 GB is size? The browse.vc.db? That can't be directly limited with a size value but you can exclude stuff, see https://github.com/microsoft/vscode-cpptools/wiki/Optimizing-your-configuration .

**Reply by tony-p** [reply]
The `~/.cache/vscode-cpptools` folder is 20Gb

**Reply by tony-p** [reply]
If I look at the settings and start searching on `C_Cpp.max` only the max number of symbols comes up as a parameter, and is the only setting under resource management, but above I also see for example `C_Cpp.maxMemory` and `C_Cpp.max*` which indicates there should be more

Update: that was on windows not on the remote WSL, their I do see more.

**Reply by sean-mcmanus** [reply]
@tony-p The .cache folder contains the .browse.vc.db file (size dependent on what is parsed in the browse.path setting versus excluded). It might also contain ipch files, which you can control the max size of with C_Cpp.intelliSenseCacheSize. The other max settings don't control the size of the .cache folder.
